### PR TITLE
substrate-client: storage improvements

### PR DESCRIPTION
**Pull request by josepot**
While I was working on the enhancers for the `substrate-client` I realized that there are a number of issues with the current `FollowResponse['storage']` function:

- It's not handling the `operationWaitingForContinue` event.
- It's not handling the `discardedItems` property of the started response.
- We are missing a more granular and subscription-based API to consume the storage API.

This PR addresses these issues:

On the one hand we have a new `FollowResponse["storageSubscription"]` api, which allows us to listen to all the relevant events. These could be useful in situations like the following: imagine that we need to query 30 different values for the same storage entry, like 30 validators... If we were to use 30 different `value` queries, that would require 30 operation slots and we are only guaranteed to have 16 available, so perhaps it could make sense to query all the descendant values in one operation and to filter out all the non-needed items. Doing that with the current Promise based API is quite inconvenient and inefficient. However, doing it with the new subscription-based API would allow us to do the same using much less memory and we could stop the operation once we have all the data that we need, without having to wait until the `operationStorageDone` event. Also, it would allow us to handle the possible discarded items more efficiently.

On the other hand: ~the API for the `FollowResponse["storage"]` function remains untouched. However, if there are any `discardedItems` it will immediately reject with an `OperationLimitError` error. Funnily enough, its internal implementation uses  the new `storageSubscription` function 🙂 .~

## EDIT:

After having put a lot of thought into this, and after having tried the pros and cons of the 2 different APIs with the downstream libraries, I'm convinced that what I mentioned in [this comment](https://github.com/paritytech/polkadot-api/pull/98#issuecomment-1732551092) is the way to go. In a nutshell:

- For advanced use cases now we will have `storageSubscription` which allows to pass various storage operations at once, while listening to the results that come over time. This is useful for things like batching many related queries in one, while resolving the ones that can be resolved as their data is already available.
- For the most typical and straightforward use cases: now the `storage` API only takes one operation at the time, it has better typings and it doesn't have to do a lot of complex magic behind the scenes.

The previous `storage` API was doing way too many things at once and as a result it was not useful for advanced use cases and it was too cumbersome for the typical and simple use cases. So, this PR amends that mistake.

### CODE REVIEW

**Comment by josepot** [comment]
I'm having second thoughts on this... I'm starting to realize that `FollowResponse["storage"]` does too much for a low-level library like this one, and that can be quite problematic.

The main reason being that the enhancer that handles the retries on `OperattionLimitError`s is going to be insanely complicated if we try to keep this API, because currently `FollowResponse["storage"]` does too much behind the scenes, which shouldn't be the case for a library that it's just meant to be a non opinionated means for consuming the new JSON RPC API.

With that in mind, I think that the best API for this library to make storage requests is:
- On the one hand: the `storageSubscription` function that's included in this PR
- On the other hand: a storage function that only let's one make one operation at a time.

File: packages/substrate-client/src/chainhead/storage-subscription.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  ) =>
+  (
+    hash: string,
+    inputs: Array<StorageItemInput>,
+    childTrie: string | null,
+    onItems: (items: Array<StorageItemResponse>) => void,
+    onError: (e: Error) => void,
+    onDone: () => void,
+    onDiscartedItems: (nDiscarted: number) => void,
+  ): (() => void) => {
```
Comment (ryanleecode) [inline packages/substrate-client/src/chainhead/storage-subscription.ts:37]:
```suggestion
  ): FollowResponse["storageSubscription"] =>
  (hash, inputs, childTrie, onItems, onError, onDone, onDiscardedItems) => {
```

File: packages/substrate-client/src/chainhead/public-types.ts [hunk 2]
```diff
@@ -67,20 +69,22 @@ export interface FollowResponse {
...
+    abortSignal?: AbortSignal | undefined,
+  ) => Promise<StorageResult<Type>>
+  storageSubscription: (
+    hash: string,
+    inputs: Array<StorageItemInput>,
+    childTrie: string | null,
+    onItems: (items: Array<StorageItemResponse>) => void,
+    onError: (e: Error) => void,
+    onDone: () => void,
+    onDiscartedItems: (nDiscarted: number) => void,
```
Comment (ryanleecode) [inline packages/substrate-client/src/chainhead/public-types.ts:86]:
```suggestion
    onDiscardedItems: (nDiscarded: number) => void,
```

**Review by kratico — APPROVED** [review]
LGTM

File: packages/substrate-client/src/chainhead/public-types.ts (hunk 2 above)
Comment (josepot) [reply packages/substrate-client/src/chainhead/public-types.ts:86]: thanks!

**Review by ryanleecode — APPROVED** [review]

**Pull request merged**